charmhub_lp_tools/charm_project.py

```diff
@@ -812,6 +819,108 @@ def _do_authorization(self, recipe: TypeLPObject) -> None:
+        builds = []
+        for recipe in self._find_recipes(branches):
+
+            build = self.request_build_by_recipe(recipe, force, dry_run)
+            if build:
+                builds.append(build)
+                print(f'{self.charmhub_name}: New build requested '
+                      f'{build.web_link}')
+
+        return builds
+
+    def request_build_by_recipe(self, recipe, force, dry_run) -> object:
+        """Request a build for the recipe.
+
+        :param recipe: recipe to request the build for.
+        :param force: if True, no checks are made to detect if the build is
+                      needed.
+        :returns: the build object or None when no build was requested.
+        """
+        build = None
+        do_build = False
+        if force:
+            logger.debug('Forcing build of recipe %s', recipe)
+            do_build = True
+        elif not self.is_build_valid(recipe.builds[0]):
+            logger.debug('Build %s is no valid, so request new build',
+                         recipe.builds[0])
+            do_build = True
+
+        if do_build:
+            if dry_run:
+                print(f'{self.charmhub_name}: New build needed (dry-run)')
+            else:
+                # the recipe build channels need to be passed when requesting
+                # the build, otherwise the build is made without the overrides
+                # that the recipe has defined.
+                build = recipe.requestBuilds(
+                    channels=recipe.auto_build_channels
+                )
+        else:
+            logger.debug('Not required to build: %s', recipe)
+
+        return build
+
+    def is_build_valid(self, build) -> bool:
+        """Determine if the build is valid.
+
+        A valid build a build that meets the following criteria:
+        - it's associated recipe has the attribute can_upload_to_store set to
+          True, but the build has no store_upload_revision set, and the build
+          is not in progress (states: currently build, uploading build or needs
+          building)
+        - the build state is in any of: 'Failed to build', 'Failed to upload'.
+        - the associated recipe is stale.
+        """
+        logger.debug(('Recipe can_upload_to_store %s , '
+                      'store_upload_revision %s, build state %s, is stale %s'),
+                     build.recipe.can_upload_to_store,
+                     build.store_upload_revision,
+                     build.buildstate, build.recipe.is_stale)
+        if (build.recipe.can_upload_to_store and
+                not build.store_upload_revision and
+                build.buildstate not in [CURRENTLY_BUILDING,
+                                         UPLOADING_BUILD,
+                                         NEEDS_BUILDING]):
+            return False
+        elif build.buildstate in [FAILED_TO_BUILD, FAILED_TO_UPLOAD]:
+            return False
+        elif build.recipe.is_stale:
+            return False
```

Comment:
This is a big logical statement? :)

```python
return (not (build.recipe ... and not build.store_upload_revision and build.buildstate not ..)
        and build.buildstate not in [...]
        and not build.recipe.is_stale)
```

At least I think it is?

(Note includes the `return True` below)

Reply:
it is ... :D let refactor it a little bit

Reply:
> This is a big logical statement? :)

I was giving to this some more thought, and the reasoning for using this chain of if-elif-elif is that each condition captures the criteria described in the docstring, so it's easier to understand why each condition is there, and easier to extend with new criteria that we may consider as "not a valid build", for example as we start seeing build per arch this method will fall short to detect when a new build is needed.

Reply:
That sounds reasonable; I'd have probably then split them into separate if statements, rather than elif each branch (which is *exactly* the same, I know), but again, it's just preferences; the code works the same way :)